### PR TITLE
Implement textDocument/documentHighlight

### LINKED ISSUE #208: Implement DocumentHighlight
`textDocument/documentHighlight` highlights related identifiers, e.g. referring to the same variable. This could probably be implemented together with #16.

### PR DESCRIPTION
**Pull request by themkat**
Really missed this in the Kotlin language server after programming Rust (using rust-analyzer) for a while. Having the same variable highlighted is super useful 🙂 

Fixes #208 

## Basic info
Think I've covered most of the weird edge cases with the unit tests. Also tested manually using Emacs (lsp-mode). One example of how it looks with a global variable being highlighted (the blue'ish):
image

...and with arguments being highlighted:
image

...and a class highlighted:
image

I think you get the point 🙂

Tried to re-use as much existing functionality as possible to make the code small and simple. Feel free to suggest any improvements if you find any. Know we have a bit different coding styles and stuff 🙂 

## Possible edge case
One weird thing I found: Hovering builtin symbols (e.g, `println` and `Exception`) and comments seems to just mark the containing function (or the nearest class or similar symbol it is commenting). I did not find this to be much of an inconvenience, so if you don't hate it either I suggest we just keep it. Unless you have a possible solution, then I'm all ears 🙂

Looks like this:
image

..and

image

## What is not supported (possible future work?)
- Highlighting an imported symbol from the import list and in the file. Currently nothing happens when you hover the import list. If you hover the symbol in the file, it only highlights the occurences that are not in the import-list. Unsure how to solve this.

Hope you are not tired of PRs from me 😆

### CODE REVIEW

**Comment by fwcd** [comment]
Wow, this looks really useful, thanks! I'm actually a bit surprised, we didn't have this before, AFAIK VSCode performs a slightly less intelligent/non-semantic variant of this highlighting already.

Will review once I find some time, I absolutely appreciate your work on this project though, very happy to see people interested in this project!

File: server/src/main/kotlin/org/javacs/kt/rename/Rename.kt [hunk 1]
```diff
@@ -28,7 +28,7 @@ fun renameSymbol(file: CompiledFile, cursor: Int, sp: SourcePath, newName: Strin
     }
 }
 
-private fun findDeclaration(file: CompiledFile, cursor: Int): Pair<KtNamedDeclaration, Location>? {
+fun findDeclaration(file: CompiledFile, cursor: Int): Pair<KtNamedDeclaration, Location>? {
```
Comment (fwcd) [inline server/src/main/kotlin/org/javacs/kt/rename/Rename.kt:31]: If we use this function elsewhere, would it make sense to move it into a utility module? Or directly into `CompiledFile` as a method?

File: server/src/main/kotlin/org/javacs/kt/highlight/DocumentHighlight.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        listOf(DocumentHighlight(declarationLocation.range, DocumentHighlightKind.Text))
+    } else {
+        emptyList()
+    } + references.map { DocumentHighlight(it, DocumentHighlightKind.Text) }
+}
+
+private fun findDeclarationCursorSite(
+        file: CompiledFile,
+        cursor: Int
+): Pair<KtNamedDeclaration, Location>? {
```
Comment (fwcd) [inline server/src/main/kotlin/org/javacs/kt/highlight/DocumentHighlight.kt:30]: If we move `findDeclaration` into `CompiledFile`, it might look nice to make this a private extension method on `CompiledFile`.

**Comment by fwcd** [comment]
Looks good, thank you very much!
